controllers/vaultsecret_controller.go

```diff
@@ -0,0 +1,226 @@
+	if nextSchedule > 0 {
+		return r.ManageSuccessWithRequeue(ctx, instance, nextSchedule)
+	} else {
+		return r.ManageSuccessWithRequeue(ctx, instance, time.Second)
+	}
+
+}
+
+func (r *VaultSecretReconciler) mergeKVSecret(mergedMap map[string]interface{}, kvSecret redhatcopv1alpha1.KVSecret, data map[string]interface{}) {
+	mergedMap[kvSecret.Name] = data
+}
+
+func (r *VaultSecretReconciler) formatK8sSecret(instance *redhatcopv1alpha1.VaultSecret, data interface{}) (*corev1.Secret, error) {
+
+	stringData := make(map[string]string)
+	for k, v := range instance.Spec.TemplatizedK8sSecret.StringData {
+		tpl, err := template.NewTemplate(&template.NewTemplateInput{
+			Contents: v,
+			Data:     data,
+		})
+
+		if err != nil {
+			r.Log.Error(err, "unable to create template", "instance", instance)
+			return nil, err
+		}
+
+		rslt, err := tpl.Execute()
+
+		if err != nil {
+			r.Log.Error(err, "unable to execute template", "instance", instance)
+			return nil, err
+		}
+		stringData[k] = string(rslt.Output)
+	}
+
+	k8sSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        instance.Spec.TemplatizedK8sSecret.Name,
+			Namespace:   instance.Namespace,
+			Annotations: instance.Spec.TemplatizedK8sSecret.Annotations,
+		},
+		StringData: stringData,
+		Type:       corev1.SecretType(instance.Spec.TemplatizedK8sSecret.Type),
+	}
+
+	return k8sSecret, nil
+}
+
+func (r *VaultSecretReconciler) manageReconcileLogic(ctx context.Context, instance *redhatcopv1alpha1.VaultSecret) error {
+
+	mergedMap := make(map[string]interface{})
+	for _, kvSecret := range instance.Spec.KVSecrets {
+		vaultClient, err := kvSecret.Authentication.GetVaultClient(ctx, instance.Namespace)
+		if err != nil {
+			r.Log.Error(err, "unable to create vault client", "instance", instance)
+			return err
+		}
+
+		ctx = context.WithValue(ctx, "vaultClient", vaultClient)
+		vaultEndpoint := vaultutils.NewVaultEndpoint(instance)
+		data, _, err := vaultEndpoint.Read(ctx)
+		if err != nil {
+			r.Log.Error(err, "unable to read Vault Secret", "instance", instance)
+			return err
+		}
+
+		r.mergeKVSecret(mergedMap, kvSecret, data)
+	}
+
+	k8sSecret, err := r.formatK8sSecret(instance, mergedMap)
+
+	if err != nil {
+		r.Log.Error(err, "unable to format k8s secret", "instance", instance)
+		return err
+	}
+
+	err = r.CreateOrUpdateResource(ctx, instance, instance.Namespace, k8sSecret)
+
+	if err != nil {
+		r.Log.Error(err, "unable to create or update k8s secret", "secret", k8sSecret)
+		return err
+	}
+
+	now := metav1.NewTime(time.Now())
+	instance.Status.LastVaultSecretUpdate = &now
+
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *VaultSecretReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	needsCreation := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			newSecret, ok := e.ObjectNew.DeepCopyObject().(*redhatcopv1alpha1.VaultSecret)
```

Comment:
I don't get this logic here....

Reply:
Its fairly standard. checks if the spec was infact changed of the CR